File: ADALplus_bounds.m
function [X, y, Z, S, primal, dual, iter, secs, info] = ADALplus_bounds(A, b, C, mleq, L, max_iter, tol, timelimit, U, ub_lam)
% ADAL: Alternating Direction Augmented Lagrangian method for SDP
%   
%       min <C, X> 
%   s.t.    A*X(:) <= b; 
%           A*X(:) = b;
%           X>=L, X psd
%
% A, b, C: A is m by n*n, b is m-vector, C is n by n
% mleq : number of <= constraints, we assume that they are from 1,...,mleq 
info = struct;
% some formal checking
[m, n2] = size(A); n = size(C, 1);
assert(length(b)==m, 'mismatch rows of A and length of b');
assert(n*n==n2, 'mismatch columns of A and size of C');
l_size = size(L);
assert((isempty(L) || l_size(1) == n), 'mismatch dimension on bounds L and U');
normb = norm(b); normC = norm(C(:));
tStart = tic;
secs_pp = 0;

if(nargin <= 5)
    max_iter=1000000;
end
if(nargin <= 6)
    tol=1e-5;
end
if(nargin <= 7)
    timelimit = 300000;
end

I = ones(m, 1);
poszeros = mleq+1:m;
I(poszeros) = 0;
% form A*A' (and cholesky, if not part of input)
if size(A,1) == m
  At = A';
else
  At = A; A = At';
end

%AAT = A * At + sparse(diag(I)); inefficient!!

AAT = A * At;

for i=1:mleq
    AAT(i,i) = AAT(i,i) + 1;
end
[R, p] = chol(AAT);
Rt = R';    % only for speed-up
secs = toc(tStart);
fprintf(' secs after chol:   %12.5f \n', secs);
if p>0 
    fprintf(' rows of A lin. dep. p = %4.0d, m = %4.0d\n',p,m);
    error(' nothing done');
end

% initialize \bar{X} = (X, x), dimension of X: (n) by (n)
X = zeros(n);
% slack variables of primal, dimension of x: (mleq) by (1)
% we pad x to dimension (m) by (1)
x = zeros(m, 1);         

% initialize \bar{Z} = (Z, z), dimension of Z: (n) by (n)
Z = zeros(n);
% surplus variables of dual, dimension of z: (mleq) by (1)
% we pad z to dimension (m) by (1)
z = zeros(m, 1);

% initialize S, dimension of S: (n) by (n) ... multipliers of X>=L
S = zeros(n);

% starting y
y = zeros(m, 1);

% inf_idx = find(isinf(L));
idx = find(~isinf(L));

lb_pp = -Inf;
secs_lb = 0;
model = post_processing_init(A,At,b,C,L,mleq);

done = 0;  % stopping condition not satisfied
iter = 0;  % iteration count
g = b - A*X(:) - x; % primal residue


G1 = Z - C;
g1 = z;
normG1 = norm( G1,'fro') + norm(g1);
relp = norm(g)^2/(1+normb); 
sigma = relp*(1+normC)/(normG1^2);
fprintf('starting sigma: %10.7f \n', sigma);


%SIGMA BOX
sigma_min = 1e-4;
sigma_max = 1e+5;

fprintf('    it     time     dObj          pObj         dFeas    pFeas     X>=L     compXS    sigma\n');

bound= -inf;
LB = -inf;
DB = -inf;
time_bound = 0;
tot_time_bound = 0;
time_LB = 0;
tot_time_LB = 0;
time_DB = 0;
tot_time_DB = 0;

model = post_processing_init(A,At,b,C,L,mleq);

% main loop
while done == 0 % while not done
    
    % update y 
    Mtmp = X/sigma - C + Z + S;    
    rs1 = b/sigma - A*Mtmp(:) - x/sigma - z; % -s;  
    tmp  = Rt\rs1;   
    y = R\tmp; 
    
    % determine new S
    Aty = reshape(At*y, n, n);
    T = C - Aty - Z - X/sigma + L/sigma;
    S = max(T, 0);
    
    % project W
    Aty = reshape(At*y, n, n);
    % Compute \bar{W} = (W, w), dimension of W: (n) by (n)
    W = Aty - C + X/sigma + S;
    % dimension of w: (mleq) by (1)
    w = x/sigma + y; % + s; 
    w = w(1:mleq);
    [Wp, wp, Wn, wn] = project_W(W, w);
    % computing new Z
    Z = -Wn; z(1:mleq) = -wn;
    % computing new X
    X = sigma * Wp; x(1:mleq) = sigma*wp;
    
    g = b - A*X(:) - x; % primal residue
    
    G = C - Z - Aty - S; % dual residue
    gg = - y - z; % - s;
    gg = gg(1:mleq);
    normX = norm(X(:));
    normx = norm(x);
    X1 = X - max(L, X);
    normX1 = norm(X1(:));
    err_1 = norm(g)/(1+normb);
    err_2 = (sqrt(norm(G(:))^2 + norm(gg)^2))/(1+normC);
    err_3 = (normX1)/(1+normX);
    XL = X(idx) - L(idx);
    normXL = norm(XL(:));
    XLt=XL.';
    trSX = S(idx).'*XLt(:);
    err_4 = abs(trSX)/(1+normXL + norm(S(:)));
    % fprintf( '%8.3f %8.3f  %8.3f %8.3f \n', trSX, normX, norm(S(:)), normx);
    
    err_0 = [err_1, err_2, err_3, err_4];
    primal = C(:)'*X(:); dual = b'*y + L(idx)'*S(idx);
    
    iter = iter + 1;
    
    if (mod(iter, 200)==0)
        tic;
        new_DB = post_processing(model,Z);
        if ceil(new_DB) > ceil(DB)
            fprintf('OLD BOUND: %13.5f, NEW BOUND %13.5f\n', ceil(DB), ceil(new_DB));
            DB = new_DB;
            time_DB = toc(tStart);
        end
        tot_time_DB = tot_time_DB + toc;
        
        %fprintf('------------------------ Valid bound: %13.5e @ %8.2f secs ------------------------------\n', DB, time_DB);
        
        lap = tic;
        K = norm(G(:));
        new_bound = dual - U * K;
        if new_bound > bound
            bound = new_bound;
            U = min(U, floor(abs(bound)) + 1);
            time_bound = toc(tStart);
        end
        tot_time_bound = tot_time_bound + toc(lap);
        

        fprintf('BOUND: %10.7f, U: %10.7f \n', bound, U);
        lap = tic;
        new_LB = rigorous_erro_bound(dual, At,C,b,X,y,Z,S, mleq,1,ub_lam);
        if new_LB > LB
            LB = new_LB;
            ub_lam = min(ub_lam, floor(abs(LB)) + 1);
            time_LB = toc(tStart);
        end
        fprintf('Error BOUND: %10.7f, ub_lam: %10.7f \n', LB, ub_lam);
        tot_time_LB = tot_time_LB + toc(lap);
        %fprintf('Rigorous BOUND: %10.7f \n', LB);
    end
    
    secs = toc(tStart);
    
    
    done = (max(err_0)<tol) | (iter>= max_iter) | (secs >= timelimit); %done=1 if max{rp,rD}<tol or iter>maxiter
    
    if done
        tic;
        new_DB = post_processing(model,Z);
        if ceil(new_DB) > ceil(DB)
            fprintf('OLD BOUND: %13.5f, NEW BOUND %13.5f\n', ceil(DB), ceil(new_DB));
            DB = new_DB;
            time_DB = toc(tStart);
        end
        tot_time_DB = tot_time_DB + toc;
        
        %fprintf('------------------------ Valid bound: %13.5e @ %8.2f secs ------------------------------\n', DB, time_DB);
        
        lap = tic;
        K = norm(G(:));
        new_bound = dual - U * K;
        if new_bound > bound
            bound = new_bound;
            U = min(U, floor(abs(bound)) + 1);
            time_bound = toc(tStart);
        end
        tot_time_bound = tot_time_bound + toc(lap);
        

        fprintf('BOUND: %10.7f, U: %10.7f \n', bound, U);
        lap = tic;
        new_LB = rigorous_erro_bound(dual, At,C,b,X,y,Z,S, mleq,1,ub_lam);
        if new_LB > LB
            LB = new_LB;
            ub_lam = min(ub_lam, floor(abs(LB)) + 1);
            time_LB = toc(tStart);
        end
        fprintf('Error BOUND: %10.7f, ub_lam: %10.7f \n', LB, ub_lam);
        tot_time_LB = tot_time_LB + toc(lap);
        %fprintf('Rigorous BOUND: %10.7f \n', LB);
    end

    % print something
    if (mod(iter, 200)==0) || done %modulus after division iter/10==0 -> stampa ogni 10 iterazioni
      fprintf( '%6.0d %8.2f %13.5e %13.5e %8.3f %8.3f  %8.3f %8.3f %9.6f \n', iter, ...
       secs, dual, primal, log10(err_0(2)), log10(err_0(1)), log10(err_0(3)),...
       log10(err_4), sigma);
    end
    % sigma update
    ratio = (sqrt(norm(X(:))^2 + norm(x)^2))/(sqrt(norm(Z(:))^2 + norm(z)^2));
    % weight
    w = 2^(-(iter-1)/100);
    %w = 1;
    %w = 0.02;
    sigma = (1-w)*sigma + w*max(sigma_min,min(ratio,sigma_max)) ;
    
%     if (mod(iter,200)==0) || done
%         tic;
%         new_lb_pp = post_processing(model,Z);
%         if round(new_lb_pp,2) > round(lb_pp,2) 
%             lb_pp = new_lb_pp;
%             secs_lb = toc(tStart);
%         end
%         secs_pp = secs_pp + toc;
%         
%         fprintf('------------------------ Valid bound: %13.5e @ %8.2f secs ------------------------------\n', lb_pp, secs_lb);
%     end    
    
end

if max(err_0) < tol
    fprintf(' required accuracy reached after %6.0d iterations,',iter)
else
    fprintf(' Time limit or Maximum number of iteration exceeded.')
end

fprintf('    total time: %8.2f \n', secs);

info.norm_bound = bound;
info.error_bound = LB;
info.time_bound = time_bound;
info.time_error_bound = time_LB;
info.tot_time_bound = tot_time_bound;
info.tot_time_error_bound = tot_time_LB;
info.dual_bound = DB;
info.time_DB = time_DB;
info.tot_time_DB = tot_time_DB;

fprintf('best Norm bound: %10.5e found at %8.2f\n', bound, time_bound);
fprintf('best Error bound: %10.5e found at %8.2f\n', LB, time_LB);
fprintf('best Dual bound: %10.5e found at %8.2f\n', DB, time_DB);
end

File: ADALplus_bounds.py
import numpy as np
import scipy.linalg as la
import time
from scipy.io import loadmat
from sksparse.cholmod import cholesky
import gurobi as gb
np.seterr(divide='ignore')


def dual_bound_init(AA, AAt, bb, CC, mleq, LL):
    A = AA.copy()
    At = AAt.copy()
    b = bb.copy()
    C = CC.copy()
    L = LL.copy()
    L_vec = L.reshape((-1, 1))
    sense = np.array(['>' if i != -np.inf else '=' for i in L_vec])
    L[L == -np.inf] = 0
    L_vec = L.reshape((-1, 1))

    Agurobi = At
    Agurobi[:, mleq:] = -Agurobi[:, mleq:]
    idx = Agurobi.getnnz(1) > 0
    Agurobi = Agurobi[Agurobi.getnnz(1) > 0]
    obj = np.vstack([A[0:mleq, :] * L_vec - b[0:mleq], b[mleq:] - A[mleq:, :] * L_vec])

    # Create Gurobi LP model
    model = gb.Model()
    model.setParam('outputflag', 0)
    model.ModelSense = gb.GRB.MAXIMIZE
    model.setParam('FeasibilityTol', 1e-5)
    y = model.addMVar(Agurobi.shape[1],
                      lb=[0.0 for i in range(mleq)] + [-float('inf') for i in range(Agurobi.shape[1] - mleq)],
                      obj=obj[:, 0], vtype=gb.GRB.CONTINUOUS, name="y")
    mc = model.addMConstr(Agurobi, y, sense[idx], (- C).reshape((-1, 1))[idx], name='dual')
    model.update()
    model.write('test.lp')
    res = {'model': model, 'C': C, 'L': L, 'idx': idx, 'mc': mc}
    return res


def dual_bound(model, Z):
    # lam, ev = la.eigh(Z)
    # Z = ev@ np.diag(lam) @ ev.T

    mc = model['mc']
    mc.setAttr("RHS", (Z - C).reshape((-1, 1))[model['idx']][:, 0])
    model['model'].update()
    model['model'].write('test.lp')
    model['model'].optimize()

    if model['model'].Status == gb.GRB.OPTIMAL:
        # a safe dual bound have been found
        p = model['C'] - Z
        lb = model['model'].ObjVal + np.sum(model['L'].reshape((1, -1)) @ p.reshape((-1, 1)))
    else:
        lb = -np.inf

    return lb


def error_bound(dual, Aty, C, X, S, y, mleq, mu=1.1, max_lamb_x=None):
    n = C.shape[0]
    # If no upper bound on the maximum eigenvalue of the optimal X is given,
    # then it is estimated by the current X, scaled by a factor mu
    max_lamb_x = mu * np.max(la.eigh(X)[0]) if not max_lamb_x else max_lamb_x

    # Compute a feasible Znew (which in general will not be Positive Semidefinite
    Znew = C - Aty - S
    znew = -y[:mleq]
    lb0 = dual
    pert = 0.
    lam, _ = la.eigh(Znew)
    I = np.where(lam < 0)[0]

    if I.any():
        pert += max_lamb_x * np.sum(lam[I])

    I = np.where(znew < 0)[0]
    if I.any():
        pert += max_lamb_x * np.sum(znew[I])

    lb = lb0 + pert
    return lb, max_lamb_x


def norm_bound(dual, K, U, debug=False):
    bound = dual - U * K
    if debug:
        print('Norm Bound: %d' % U)
        print('Safe dual bound: %13.4f' % bound)
    return bound


def project(X, L, U):
    # Works only with Numpy Data types
    if X.shape == ():
        return L if X < L else (U if X > U else X)
    idx_l = X < L
    idx_u = X > U
    X[idx_l] = L[idx_l]
    X[idx_u] = U[idx_u]
    return X

def ADMM_bounds(A, b, C, mleq, L, sigma=1., options={}):
    # Initialization
    tstart = time.time()
    # Read option from dict
    tol = 1e-5 if 'tolerance' not in options else options['tolerance']
    max_iter = 1000 if 'max_iter' not in options else options['max_iter']
    timelimit = 3600 if 'timelimit' not in options else options['timelimit']
    print_it = 100 if 'print_it' not in options else options['print_it']
    debug = True if 'debug' not in options else options['debug']
    num_iter = 1
    done = False
    result = {}

    # initialization of sigma box
    t_min = np.float64(1e-4)
    t_max = np.float64(1e+7)

    m, n2 = A.shape
    n = int(np.sqrt(n2))

    assert L is None or L.shape[0] == n, 'mismatch dimension on bounds L'

    At, AAT = A.T, A @ A.T

    AAT_lil = AAT.tolil()
    # Add slack variables to mleq constraints
    AAT_lil.setdiag(AAT_lil.diagonal() + np.array([1 if i < mleq else 0 for i in range(m)]))
    AAT = AAT_lil.tocsc()

    factor = cholesky(AAT)
    secs = time.time() - tstart

    if debug:
        print('Cholesky factorization completed after: %12.5f' % secs)

    # initialize primal variable
    Y = np.zeros((n, n))
    # initialize dual variable
    Z = np.zeros((n, n))
    # initialize S, dimension of S: (n) by (n) ... multipliers of X>=L
    S = np.zeros((n, n))

    # Slack variables of the primal, dimension of x: (mleq) by (1)
    # we pad x to dimension (m) by (1)
    x = np.zeros((m, 1))
    # Surplus variables of the primal, dimension of z: (mleq) by (1)
    # we pad z to dimension (m) by (1)
    z = np.zeros((m, 1))

    idx = L!=-np.inf

    normb, normC = la.norm(b), la.norm(C)

    # Needed to initialize the LP for dual bound safe bound
    # model = dual_bound_init(A, At, b, C, mleq, L)

    if debug:
        print(' it     secs       dual        primal       dFeas    pFeas     X>=L     compXS   sigma  ')

    while not done:
        # weight for sigma update
        w = np.power(2, -(num_iter - 1) / 100)

        # given Y, Z, S and sigma, solve for y
        M_tmp = Y/sigma - C + Z + S
        rhs = b/sigma - A*M_tmp.reshape((-1, 1)) - x/sigma - z
        y = factor(rhs)
        Aty = (At * y).reshape((n, n))

        S = C - Aty - Z - Y / sigma + L / sigma
        S[S < 0] = 0

        M = Aty - C + Y / sigma + S

        M1 = x / sigma + y
        M1 = M1[:mleq]

        lam, ev = la.eigh(M)
        I = np.where(lam > 0)[0]
        j = len(I)
        if j < n / 2:
            evp = np.zeros((n, j))
            for r in range(j):
                ic = I[r]
                evp[:, r] = ev[:, ic] * np.sqrt(lam[ic])
            if j == 0:
                evp = np.zeros((n, 1))
            Mp = evp @ evp.T
            Mn = M - Mp
        else:
            I = np.where(lam < 0)[0]
            j = len(I)  # should be <= n/2
            evn = np.zeros((n, j))
            for r in range(j):
                ic = I[r]
                evn[:, r] = ev[:, ic] * np.sqrt(-lam[ic])
            if j == 0:
                evn = np.zeros((n, 1))
            Mn = -evn @ (evn.T)
            Mp = M - Mn

        # Project the diagonal part of M
        mp = M1.copy()
        mp[mp < 0] = 0
        mn = M1 - mp

        Z = -Mn
        z[:mleq] = -mn
        X = sigma * Mp
        x[:mleq] = sigma * mp

        # Y update
        Y = X

        g = b - A * Y.reshape((-1, 1)) - x
        G = C - Aty - Z - S
        gg = - y - z
        gg = gg[:mleq]

        normX = la.norm(X)

        err_d = (la.norm(G) + la.norm(gg))/(1 + normC)
        dual = ((b.T @ y) + np.sum(L[idx].reshape((1, -1)) @ S[idx].reshape((-1, 1)))).item()
        err_p = la.norm(g)/(1 + normb)
        primal = np.sum(C.reshape((1, -1)) @ Y.reshape((-1, 1)))
        rel_err_p, rel_err_d = err_p / (1 + normb), err_d / (1 + normC)
        err_X_L = la.norm(X - np.maximum(L, X))/normX
        XL = X[idx] - L[idx]
        compXS = np.abs(np.sum(S[idx].reshape((1, -1)) @ XL.reshape((-1, 1))))/ (1+la.norm(XL) + la.norm(S))
        secs = time.time() - tstart
        num_iter = num_iter + 1

        # Printing
        if (num_iter % print_it) == 0:
            if debug:
                print('%3.0d %8.2f %13.5e %13.5e %8.3f %8.3f  %8.3f %8.3f  %9.6f' %
                      (num_iter, secs, dual, primal, np.log10(rel_err_d), np.log10(rel_err_p), np.log10(err_X_L),
                       np.log10(compXS), sigma))

        # Stopping criteria
        if np.max([rel_err_d, rel_err_p, err_X_L, compXS]) < tol or num_iter > max_iter or secs > timelimit:
            if debug:
                print('%3.0d %8.2f %13.5e %13.5e %8.3f %8.3f  %8.3f %8.3f  %9.6f' %
                      (num_iter, secs, dual, primal, np.log10(rel_err_d), np.log10(rel_err_p), np.log10(err_X_L),
                       np.log10(compXS), sigma))

            if debug:
                print('total time: %10.3f' % secs)
            if num_iter > max_iter:
                if debug:
                    print('max outer iterations reached.')
            if secs > timelimit:
                if debug:
                    print('Time limit exceeded')

            done = 1

        ratio = (la.norm(X) + la.norm(x)) / (la.norm(Z) + la.norm(z))
        sigma = (1 - w) * sigma + w * project(ratio, t_min, t_max)
    # Calls to safe bounding procedures
    # db = dual_bound(model, Z)
    # eb = error_bound(dual, Aty, C, X, S, y, mleq)
    # nb = norm_bound(dual, (la.norm(G) + la.norm(gg)), U)
    return result

# Example usage
h = loadmat(os.path.join('instances', 'test_adal.mat'))
A, b, C, mleq, L = h['A'], h['b'], h['C'], h['mleq'].sum(), h['L']
ADMM_bounds(A, b, C, mleq, L, sigma=50., options={'tolerance' : 1e-6})